Update cart item_data output

ralucaStan opened the pull request:
Fixes WOOPLUG-5810

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->
Following #61437 the separator for the item_data was moved outside of the `span` for the `item_data` name.

- This PR puts back the separator inside the name `span` for the Cart, Checkout and the Mini-Cart block. Pointed out internally on Slack
- Refactors the logic in the iAPI Mini-Cart around `CartItemDataAttribute`. Coming back to it, I struggled each time to remember the logic and the data structure, and faced TS errors because we were handling the visibility calculation via a hidden flag.


<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->





<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add a simple product and a variable product to the cart
2. Notice the cart items in the Cart, iAPI Mini-Cart, and Checkout
3. Check that the `:` separators are included in the variation name span

4. Add custom `item_data` to products, use snippet from https://github.com/woocommerce/woocommerce/pull/61437
5. Notice the cart items in the Cart, iAPI Mini-Cart, and Checkout
6. Check that the `:` separators are included in the variation name span

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

ralucaStan committed:
Add back the : symbol to the item data name

ralucaStan committed:
Refactor item_data logic in the iAPI store:
- remove hidden flag from item data and create getter for visibility
- put back the : symbol in the item_data name span element

ralucaStan committed:
changelog

ralucaStan committed:
Put back comment

ralucaStan committed:
Merge branch 'trunk' into fix/cart-item-data-output

ralucaStan committed:
Merge branch 'trunk' into fix/cart-item-data-output

mikejolley reviewed (approved):
Checked variations and simple products in mini-cart/cart with and without the itemData snippet. Confirmed the `:` was correctly placed. Thank you 👍🏻

PanosSynetos reviewed (approved):
I manually tested this with Product Bundles, and it works as expected

<img width="2019" height="1496" alt="Image" src="https://github.com/user-attachments/assets/48a036ff-c2e5-44f5-9f83-c5adff501579" />

ralucaStan committed:
Merge branch 'trunk' into fix/cart-item-data-output

SantosGuillamot commented on plugins/woocommerce/client/blocks/assets/js/blocks/mini-cart/iapi-frontend.ts:
```diff
@@ -832,14 +834,35 @@ const { state: cartItemState } = store(
 			},
 
+			get cartItemDataAttrHidden(): boolean {
+				return cartItemState.cartItemDataAttr === null;
```
This does not take into account that when `hidden` inside `item_data` is '1', it should hide the element. The line removed [here](https://github.com/woocommerce/woocommerce/pull/61839/files#diff-e5321b2c65fe5066069deccc33f89005ef1a93a664f49f9791960613fbf35e4fL835). This means that if I pass it as in the following example, it isn't hidden:

```js
function test_get_item_data( $cart_item_data, $cart_item ) {
	$cart_item_data[] = array(
		'key'    => 'Test Name',
		'value'  => 'Test Value',
		'hidden' => false,
	);

	$cart_item_data[] = array(
		'key'    => 'Test Name 2',
		'value'  => 'Test Value 2',
		'hidden' => true,
	);

	return $cart_item_data;
}
add_filter( 'woocommerce_get_item_data', 'test_get_item_data', 10, 2 );
```

We can compare it with how it works in the Cart or in `trunk` right now.

ralucaStan replied:
Thank you for catching that. I wasn't aware that `hidden` can be passed via the filter, my bad here, I should have checked. Why did you only checked for `dataItemAttr.hidden === '1' `?

ralucaStan replied:
I have an answer here https://github.com/woocommerce/woocommerce/pull/61839#discussion_r2519152061. I'll update my code based on the new information

SantosGuillamot commented on plugins/woocommerce/client/blocks/assets/js/blocks/mini-cart/iapi-frontend.ts:
```diff
@@ -832,14 +834,35 @@ const { state: cartItemState } = store(
+				}
+
+				return dataAttr.name + ':';
+			},
```
I wonder if it would make sense to move this logic directly to the `cartItemDataAttr` logic. I find it weird that we need derived state for another derived state property just to handle the case where it doesn't exist. I mean adding something like this [here](https://github.com/woocommerce/woocommerce/pull/61839/files#diff-e5321b2c65fe5066069deccc33f89005ef1a93a664f49f9791960613fbf35e4fR829-R837):

```js
return {
	name: dataItemAttr ? nameTxt.value + ':' : '',
	value: valueTxt.value || '',
	className: dataItemAttr ? `wc-block-components-product-details__${ nameTxt.value
		.replace( /([a-z])([A-Z])/g, '$1-$2' )
		.replace( /<[^>]*>/g, '' )
		.replace( /[\s_&]+/g, '-' )
		.toLowerCase() }` : '',
	hidden: ( ! dataItemAttr || dataItemAttr.hidden === '1' ) ? true : false,
};
```

I haven't tested the code, but just to give an example.

ralucaStan replied:
> I wonder if it would make sense to move this logic directly to the `cartItemDataAttr` logic.

I've addressed this in https://github.com/woocommerce/woocommerce/pull/61839/commits/7fcfd589c3026b341ae0b04205c33201957c5eb5, along with adding back the hidden key. I've kept the cartItemDataAttr as the source of truth, and eliminated the need for separate getters. 

I'd appreciate another review tomorrow, as the Feature freeze for 9.4 is on Monday. Thank you!

ralucaStan committed:
Account for hidden key in cart item_data & remove derived data from cartItemDataAttr

This commit keeps one source of truth for the cartItemDataAttr, preventing the need to have separate getters for name, value and class. It also adds back the hidden key to the cartItemDataAttr and ItemData key, as the hidden flag can be passed via filters

ralucaStan committed:
Merge branch 'fix/cart-item-data-output' of github.com:woocommerce/woocommerce into fix/cart-item-data-output

ralucaStan committed:
Account for '1' as value for the hidden key coming from Store API after filters

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/blocks/mini-cart/iapi-frontend.ts:
```diff
@@ -824,22 +827,31 @@ const { state: cartItemState } = store(
 				valueTxt.innerHTML = rawValue;
 
+				const processedName = nameTxt.value ? nameTxt.value + ':' : '';
+
 				return {
-					name: nameTxt.value,
+					name: processedName,
 					value: valueTxt.value,
 					className: `wc-block-components-product-details__${ nameTxt.value
 						.replace( /([a-z])([A-Z])/g, '$1-$2' )
 						.replace( /<[^>]*>/g, '' )
 						.replace( /[\s_&]+/g, '-' )
 						.toLowerCase() }`,
-					hidden: dataItemAttr.hidden === '1' ? true : false,
+					hidden: !!dataItemAttr.hidden,
 				};
```
I applied the suggestions, but I couldn't identify a path where hidden will come with value 1 (number) from Store API, so I will only check for `true` or `1`

ralucaStan committed:
Account for 'true' and 1 when checking the hidden value;

I could see 'true' as a value coming from Store API

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/blocks/mini-cart/iapi-frontend.ts:
```diff
@@ -824,22 +827,32 @@ const { state: cartItemState } = store(
 						.replace( /[\s_&]+/g, '-' )
 						.toLowerCase() }`,
-					hidden: dataItemAttr.hidden === '1' ? true : false,
+					hidden: hiddenValue === true || hiddenValue === '1',
```
I think we can disregard values like 'yes' coming from extensions. It's up to extensions to use proper values.

SantosGuillamot reviewed (approved):
LGTM 🙂
